File: examples/go/sanitizer_test/handlers.go
// Go sanitizer-fixture — parallel handlers per sink family.
package sanitizer_test

import (
	"crypto/subtle"
	"database/sql"
	"fmt"
	"html/template"
	"net/http"
	"net/url"
	"os/exec"
)

// --- Command injection ----------------------------------------------------

func CmdRaw(w http.ResponseWriter, r *http.Request) {
	cmd := r.URL.Query().Get("cmd")
	out, _ := exec.Command("sh", "-c", "ping "+cmd).Output()
	w.Write(out)
}

func CmdSafe(w http.ResponseWriter, r *http.Request) {
	cmd := r.URL.Query().Get("cmd")
	// argv-form without a shell wrapper: `cmd` becomes a single arg,
	// not a shell command. This is Go's canonical cmdi mitigation.
	out, _ := exec.Command("ping", cmd).Output()
	w.Write(out)
}

// --- SQL injection --------------------------------------------------------

func SqlRaw(db *sql.DB, userID string) *sql.Rows {
	rows, _ := db.Query("SELECT * FROM users WHERE id = '" + userID + "'")
	return rows
}

func SqlSafe(db *sql.DB, userID string) *sql.Rows {
	stmt, _ := db.Prepare("SELECT * FROM users WHERE id = $1")
	rows, _ := stmt.Query(userID)
	return rows
}

// --- XSS ------------------------------------------------------------------

func XssRaw(w http.ResponseWriter, r *http.Request) {
	name := r.URL.Query().Get("name")
	fmt.Fprintf(w, "<p>Hello, %s</p>", name)
}

func XssSafe(w http.ResponseWriter, r *http.Request) {
	name := r.URL.Query().Get("name")
	safe := template.HTMLEscapeString(name)
	fmt.Fprintf(w, "<p>Hello, %s</p>", safe)
}

// --- Open redirect --------------------------------------------------------

func RedirectSafe(w http.ResponseWriter, r *http.Request) {
	target := r.URL.Query().Get("to")
	safe := url.QueryEscape(target)
	http.Redirect(w, r, "/next?to="+safe, http.StatusFound)
}

// --- Timing attack --------------------------------------------------------

func TokenEqRaw(given, expected string) bool {
	return given == expected
}

func TokenEqSafe(given, expected string) bool {
	return subtle.ConstantTimeCompare([]byte(given), []byte(expected)) == 1
}

File: examples/c/sanitizer_test/handlers.c
/* C sanitizer-fixture — parallel handlers per sink family. */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sqlite3.h>

/* --- SQL injection -------------------------------------------------- */

void sql_raw(sqlite3 *db, const char *user_id) {
    sqlite3_stmt *stmt;
    char q[256];
    snprintf(q, sizeof q, "SELECT * FROM users WHERE id = '%s'", user_id);
    sqlite3_prepare_v2(db, q, -1, &stmt, NULL);
}

void sql_safe(sqlite3 *db, const char *user_id) {
    sqlite3_stmt *stmt;
    sqlite3_prepare_v2(db, "SELECT * FROM users WHERE id = ?", -1, &stmt, NULL);
    sqlite3_bind_text(stmt, 1, user_id, -1, SQLITE_TRANSIENT);
}

/* --- Path buffer safety --------------------------------------------- */

void path_raw(const char *name, char *out, size_t outsz) {
    strcat(out, name);  /* unbounded — caller must have sized out. */
}

void path_safe(const char *name, char *out, size_t outsz) {
    strlcpy(out, name, outsz);  /* bounded by outsz. */
}
